ability of adding new advert position from project-base

vitek-rostislav opened the pull request:
| Q             | A
| ------------- | ---
|Description, reason for the PR| Project developer must be able to add new advert position easily
|New feature| Yes
|BC breaks| Yes
|Fixes issues| https://github.com/shopsys/shopsys/issues/392
|Standards and tests pass| Yes
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

PetrHeinz commented on packages/framework/src/Model/Advert/AdvertPositionRegistry.php:
```diff
@@ -0,0 +1,32 @@
+    const POSITION_HEADER = 'header';
+    const POSITION_FOOTER = 'footer';
+    const POSITION_PRODUCT_LIST = 'productList';
+    const POSITION_LEFT_SIDEBAR = 'leftSidebar';
```
Personally, I wouldn't use the constants at all

PetrHeinz replied:
Personally, I wouldn't use the constants at all.

PetrHeinz commented on packages/framework/src/Model/Advert/Exception/AdvertPositionNotKnownException.php:
```diff
@@ -0,0 +1,19 @@
+
+class AdvertPositionNotKnownException extends Exception implements AdvertException
+{
+    public function __construct(string $positionName, array $knownPositionsNames)
```
You've omitted all annotations. Why?

PetrHeinz commented on packages/framework/src/Model/Advert/Exception/AdvertPositionNotKnownException.php:
```diff
@@ -0,0 +1,19 @@
+    public function __construct(string $positionName, array $knownPositionsNames)
+    {
+        $message = sprintf(
+            'Unknown advert position name: "%s". These are known names: %s.',
```
Quotes are missing around the second placeholder.

PetrHeinz commented on packages/framework/src/Model/Advert/AdvertFacade.php:
```diff
@@ -69,6 +77,7 @@ public function getById($advertId)
      */
     public function findRandomAdvertByPositionOnCurrentDomain($positionName)
     {
+        $this->advertPositionRegistry->assertPositionNameIsKnown($positionName);
```
We usually separate return statement with a blank line. In the past, this was (still is?) a part of the coding standards.

PetrHeinz commented on packages/framework/src/Model/Advert/AdvertPositionRegistry.php:
```diff
@@ -0,0 +1,32 @@
+    const POSITION_PRODUCT_LIST = 'productList';
+    const POSITION_LEFT_SIDEBAR = 'leftSidebar';
+
+    public function getTranslatedPositionDescriptionsIndexedByPositionName(): array
```
You've omitted all annotations. Why?

PetrHeinz replied:
Also, the method's name is a bit mouthful. Wouldn't `getAllLabelsIndexedByNames` suffice?

vitek-rostislav replied:
I though that the typehint is sufficient but I forgot that it is not specific enough, so I added the `string[]` annotation :slightly_smiling_face: 

You are right about the name :laughing: I changed it :+1:

PetrHeinz commented on packages/framework/src/Model/Advert/AdvertPositionRegistry.php:
```diff
@@ -0,0 +1,32 @@
+        ];
+    }
+
+    public function assertPositionNameIsKnown(string $positionName)
```
`: void`?

PetrHeinz commented on packages/framework/src/Model/Advert/Exception/AdvertPositionNotKnownException.php:
```diff
@@ -0,0 +1,19 @@
+
+class AdvertPositionNotKnownException extends Exception implements AdvertException
+{
+    public function __construct(string $positionName, array $knownPositionsNames)
```
In Exceptions' constructors, we always add a optional last parameter `Exception $previous = null`. In the past, this was (still is?) a part of the coding standards.

vitek-rostislav replied:
True, fixed :+1:

PetrHeinz commented on packages/framework/src/Model/Advert/Exception/AdvertPositionNotKnownException.php:
```diff
@@ -0,0 +1,19 @@
+    public function __construct(string $positionName, array $knownPositionsNames)
+    {
+        $message = sprintf(
+            'Unknown advert position name: "%s". These are known names: %s.',
```
Quotes are missing around the second placeholder.

Also, the formulation is a bit dry. How about `Unknown advert position name "%s". Known names are "%s".`?

vitek-rostislav replied:
quotes are missing because I want to quote each name separately :slightly_smiling_face: I will change the formulation though :+1:

PetrHeinz commented on project-base/src/Shopsys/ShopBundle/Controller/Front/ProductController.php:
```diff
@@ -175,13 +174,7 @@ public function listByCategoryAction(Request $request, $id)
-
-            return $this->render('@ShopsysShop/Front/Content/Product/list.html.twig', $viewParameters);
-        }
+        return $this->render('@ShopsysShop/Front/Content/Product/ajaxList.html.twig', $viewParameters);
```
Please, can you explain the thought behind this change? I'm a bit baffled.

vitek-rostislav replied:
Oh, it was by mistake, thanks for noticing that :+1: I though that the template was the same but now I see it was not :man_facepalming:

vitek-rostislav replied:
fixed :innocent:

vitek-rostislav commented:
Hi @PetrHeinz, thx for the review, I fixed everything and added entries into the `changelog` and `upgrade` files and added the cookbook. Unfortunately, I performed some history changes locally before your review was done so I then push forced it along with the fixes. I hope it wont make the review much harder for you :innocent:

PetrHeinz commented on docs/cookbook/adding-a-new-advert-position.md:
````diff
@@ -0,0 +1,51 @@
+    ```yaml
+    Shopsys\ShopBundle\Model\Advert\AdvertPositionRegistry: ~
+    
+    Shopsys\FrameworkBundle\Model\Advert\AdvertPositionRegistry: '@Shopsys\ShopBundle\Model\Advert\AdvertPositionRegistry'
````
I would rather recommend
```yaml
Shopsys\FrameworkBundle\Model\Advert\AdvertPositionRegistry: 
    class: Shopsys\ShopBundle\Model\Advert\AdvertPositionRegistry
```
That should work as well, right?

PetrHeinz commented on UPGRADE.md:
```diff
@@ -16,12 +16,16 @@ There is a list of all the repositories maintained by monorepo, changes in log b
 * [shopsys/product-feed-zbozi]
 
 ## [From 7.0.0-alpha4 to Unreleased]
+### [shopsys/framework]
```
Please add a blank line above for consistency

PetrHeinz commented on UPGRADE.md:
```diff
@@ -16,12 +16,16 @@ There is a list of all the repositories maintained by monorepo, changes in log b
     - update composer dependencies `composer update`
     - create Elasticsearch indexes by running `php phing elasticsearch-indexes-create`
     - export products into Elasticsearch by `php phing elasticsearch-products-export`
+- check for your usages of advert position names constants, they were removed from `Advert` class
```
Maybe we should just leave the constants in there (maybe marked as `@deprecated`?) so we won't make an unnecessary BC-break...

PetrHeinz replied:
You haven't even removed them! :D

PetrHeinz commented on docs/cookbook/adding-a-new-advert-position.md:
```diff
@@ -0,0 +1,51 @@
+# Adding a New Advert Position
+
+By default, you can place advert banner on predefined positions:
```
**an** advert position

PetrHeinz commented on docs/cookbook/adding-a-new-advert-position.md:
```diff
@@ -0,0 +1,51 @@
+
+![Advert detail in administration](img/new-advert-screen.png)
+
+If you want to add a new custom position, let us say on product detail page, you need to follow these steps:
```
**the** product detail page

PetrHeinz commented on docs/cookbook/adding-a-new-advert-position.md:
````diff
@@ -0,0 +1,51 @@
+    Shopsys\FrameworkBundle\Model\Advert\AdvertPositionRegistry: '@Shopsys\ShopBundle\Model\Advert\AdvertPositionRegistry'
+    ```
+
+3. Render the new advert on the product detail page (or wherever else you want) by adding following line into the corresponding twig template.
````
**the**  following line

PetrHeinz commented on docs/cookbook/adding-a-new-advert-position.md:
```diff
@@ -0,0 +1,51 @@
+- in a category detail (above the category name)
+
+All the positions are defined in `AdvertPositionRegistry` class 
+that provides them for displaying on adverts list in admin
```
Having every sentence on a single line helps with the diffs and further maintenance. Consider removing the line breaks.

PetrHeinz commented on docs/cookbook/adding-a-new-advert-position.md:
```diff
@@ -0,0 +1,51 @@
+
+All the positions are defined in `AdvertPositionRegistry` class 
+that provides them for displaying on adverts list in admin
+as well as for options for selectbox on advert detail administration page.
```
as well as for **the** options **in the** selectbox on **the** advert detail administration page**:**

*Sneaky little articles...*

PetrHeinz commented on docs/cookbook/adding-a-new-advert-position.md:
````diff
@@ -0,0 +1,51 @@
+3. Render the new advert on the product detail page (or wherever else you want) by adding following line into the corresponding twig template.
+    ```twig
+    {{ render(controller('ShopsysShopBundle:Front/Advert:box',{'positionName' : 'productDetail'})) }}
+    ```
````
Some kind of conclusion (maybe with an illustration screenshot?) is missing at the end.

Otherwise, the cookbook is perfect! :medal_sports:

PetrHeinz commented:
@vitek-rostislav: I've tweaked it in a couple of fixup commits as we agreed, please review and merge if it's OK.
